CORE_NAMING_GUIDELINE.md

```diff
@@ -0,0 +1,181 @@
+## Naming Rules and Guidelines
+In software development, the right naming conventions bridge the gap between human linguistic patterns and logical code structures, turning abstract concepts into meaningful expressions and algorithmic intent. By aligning our codebase with familiar linguistic constructs, we aim to foster clarity, maintainability, and a more intuitive understanding of our project. Embracing lean and clean code principles, this guideline aims at a concise yet expressive naming. While we strive for consistency, we emphasize the spirit of clarity and meaning over rigid adherence when faced with exceptions.
+
+This guideline is embedded into the larger [CORE_DEVELOPMENT_GUIDELINE.md](../blob/main/CORE_DEVELOPMENT_GUIDELINE.md) context of this project [README.md](../blob/main/README.md).
+<details> <summary>click here for the TL;DR summary</summary>
+
+## Naming Conventions: Quick Summary (TL;DR)
+
+- **General Philosophy**: Strive for clarity and intuition. When in doubt, prioritize clarity over strict adherence to these rules.
+- **Classes/Structs/Enums**: Use `UpperCase` like `Graph` or `SubGraph`.
+- **Methods/Functions/Lambdas**: Use `lowerCase` such as `start()` or `create()`.
+- **Fields**: Use `snake_case` for public fields like `is_valid`. Non-public fields start with `_`, e.g., `_initialised`.
```

Comment:
Is there any preference of AE vs BE? (Like initialized vs initialised) should it be specified in the guidelines?

Reply:
**Short**: We should neither enforce nor prefer one style over the other. Instead, we should embrace the diversity and cultural backgrounds of our contributors by allowing them to choose the language style that best represents them. @mormj may comment on the official GNU Radio policy, if there is any.

At GSI/FAIR, many of our members hail from the EU or from countries where English is the official language (like the UK & India). Consequently, we lean towards the EU's approach and BE. The EU describes this as *"the variety of English [... which] is the shared standard usage of Ireland and the United Kingdom"* [link](https://commission.europa.eu/system/files/2023-07/styleguide_english_dgt_en.pdf). This means our style broadly aligns with guides such as the OED, COD, and Macquarie.

On the other hand, GNU Radio has a robust base of support and contributions from the US. In terms of AE, primary influences include Merriam-Webster and style guides like Chicago or AP. This means that institutions like NSF, DOE, and other US labs likely adhere to these styles.

While there is overlap, BE and AE have distinctions. I have a fondness for the 'Oxford Comma' as I find it logical. However, the saying goes: "Whose bread I eat, his song I sing." In practical terms, contributors from the EU, UK, India, the Commonwealth, and most parts of the world should ideally align with the BE-OED style. In contrast, US contributors might consider adhering to the styles mandated by their funding agencies.

Reply:
Was curious about the UN and ISO standards... they largely follow BE (now codified). The ISO C++ standard does not codify this but largely use AE over BE (Bjarne worked at Bell Labs). Most of this is historically grown ... am not sure whether there is a technical gain in choosing one over the other.

Either way, people should check their spelling and grammar ...

Reply:
FWIW, I've never (knowingly) contributed to a software(-related) project before, that didn't agree on AE spelling. As soon as AE vs BE spelling crops up in function/type/concept names, a "the contributor decides" isn't good enough anymore, IMO. Consistency is more important.
As long as we work on GSI/FAIR internal software I wouldn't care too much if we choose BE spelling (though I'd use the wrong spelling first, because I'm simply used to all API using AE spelling). But this project is targeting a global & international audience and I'd prefer to follow the most common convention here. Now, I don't have hard numbers, only experience from Qt, KDE, GCC, and the C++ standard — all AE, AFAIK.

Reply:
While your examples lean towards AE in some aspects, they aren't exclusively AE but blend both BE and AE. This is primarily due to historical influences rather than a clear mandate. BE is more prevalent globally in organisations like the UN, ISO, etc., making it a more inclusive choice.
However, diving deeper into this topic is not productive. The BE vs. AE debate often leans towards cultural norms and trivial "bike-shedding" rather than technical solid necessities.

On a lighter note, if we want to adopt a universally neutral language, we should all revert to Lingua Franca (French) or, even better, go old school with Latin! 😂